pubtools/_quay/signature_handler.py

```diff
@@ -0,0 +1,596 @@
+                self.target_settings["pyxis_server"],
+                "--pyxis-krb-principal",
+                self.target_settings["iib_krb_principal"],
+            ]
+            if "iib_krb_ktfile" in self.target_settings:
+                args += ["--pyxis-krb-ktfile", self.target_settings["iib_krb_ktfile"]]
+            args += ["--signatures", json.dumps(batch)]
+
+            LOG.info("Uploading signature batch #{0}/{1}".format(i + 1, len(signature_batches)))
+
+            env_vars = {}
+            run_entrypoint(
+                ("pubtools-pyxis", "console_scripts", "pubtools-pyxis-upload-signatures"),
+                "pubtools-pyxis-upload-signature",
+                args,
+                env_vars,
+            )
+
+    def validate_radas_messages(self, claim_messages, signature_messages):
+        """
+        Check if messages received from RADAS contain any errors.
+
+        Args:
+            claim_messages ([dict]):
+                Messages sent to RADAS.
+            signature_messages ([dict]):
+                Messages received from RADAS.
+
+        Raises:
+            SigningError:
+                If RADAS messages contain errors.
+        """
+        failed_messages = 0
+        for message in signature_messages:
+            if message["errors"]:
+                failed = [m for m in claim_messages if m["request_id"] == message["request_id"]][0]
+                LOG.error(
+                    "Signing of claim message {0} failed with following errors: {1}".format(
+                        failed, message["errors"]
+                    )
+                )
+                failed_messages += 1
+
+        if failed_messages:
+            raise SigningError(
+                "Signing of {0}/{1} messages has failed".format(
+                    failed_messages, len(claim_messages)
+                )
+            )
+
+
+class ContainerSignatureHandler(SignatureHandler):
+    """Class for handling the signing of container images."""
+
+    def construct_item_claim_messages(self, push_item):
+        """
+        Construct all the signature claim messages for RADAS for one push item.
+
+        push_item (ContainerPushItem):
+            Container push item whose claim messages will be created.
+        Returns ([dict]):
+            Claim messages for a given push item.
+        """
+        LOG.info("Constructing claim messages for push item '{0}'".format(push_item))
+        claim_messages = []
+        digests = self.get_tagged_image_digests(push_item.metadata["pull_url"])
+        # each image digest needs its own signature
+        for digest in digests:
+            # each destination image reference needs its own signature
+            for repo, tags in sorted(push_item.metadata["tags"].items()):
+                for tag in tags:
+                    claim_messages += self.construct_variant_claim_messages(
+                        repo, tag, digest, [push_item.claims_signing_key]
```

Comment:
--signing-key is not mandatory for "pub push-docker". Should signing be skipped if no signing key is provided?

Reply:
@midnightercz Is it possible that push item will have no signing key?